doc/man/gnunet-directory.1
.TH gnunet-directory "1" "23 Dec 2006" "GNUnet"
.SH NAME
gnunet\-directory \- display directories and show or delete lists of encountered file identifiers

.SH SYNOPSIS
.B gnunet\-directory
[\fIOPTIONS\fR] FILENAMES
.SH DESCRIPTION
.PP
gnunet\-directory lists the contents of a GNUnet directory.  It also can be used to manipulate the file identifier database which is used by GNUnet for building directories.
.PP
gnunet\-directory will always list the contents of the GNUnet directories that are passed as filenames.
.PP
Manipulating the file identifier database is done by passing additional options to gnunet\-directory.  Note that by default GNUnet does not build the file identifier database and the database will thus always be empty.  You need to run gnunet\-directory with the -\t option to enable tracking of file identifiers.  The reason is that storing file identifiers in plaintext in the database can compromise your privacy if your machine should fall under the control of an adversary.

.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
use config file (defaults: ~/.gnunet/gnunet.conf)
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-k, \fB\-\-kill\fR
delete all entries from the file identifier database and stop tracking file identifiers
.TP
\fB\-l, \fB\-\-list\fR
display entries from the file identifier database
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
Change the loglevel.  Possible values for LOGLEVEL are NOTHING, FATAL,
ERROR, WARNING, INFO, STATUS and DEBUG.  Note that options in the 
configuration file take precedence over this option (the argument 
will be ignored in that case).

.TP
\fB\-t, \fB\-\-track\fR
start tracking file identifiers
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-V\fR, \fB\-\-verbose\fR
be verbose; prints progress information
.SH NOTES

The options are evaluated in the order \-l, \-k and then \-t.  Thus specifying \-klt will first list all identifiers that were in the database previously, then delete all of those identifiers from the database and then continue (or start to) collect identifiers.  The order in which the options are specified on the command line is not important.

A GNUnet directory is a file containing a list of GNUnet URIs and meta data.  The keys can point to files, other directories or files in namespaces.  In other words, a GNUnet directory is similar to UNIX directories.  The difference to tar and zip is that GNUnet directory does not contain the actual files, just symbolic (links), similar to directories with symbolic links in UNIX filesystems.  The benefit is that the individual files can be retrieved separately (if desired) and if some of the files are inserted to another node in GNUnet, this just increases their availability but does not produce useless duplicates (for example, it is a better idea to publish a collection of pictures or compressed sound files using a GNUnet directory instead of processing them with archivers such as tar or zip first).  Directories can contain arbitrary meta data for each file.

At the moment, directories can be created by \fBgnunet\-gtk\fP and \fBgnunet\-insert\fP.  They can point to content created by the user or content inserted by others.  Just like ordinary files, a directory can be published in a namespace.

GNUnet directories use the (unregistered) mimetype \fBapplication/gnunet\-directory\fP.  They can show up among normal search results.  The directory file can be downloaded to disk by \fBgnunet\-download\fP(1) for later processing or be handled more directly by \fBgnunet\-gtk\fP(1).

.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-gtk\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-search\fP(1), \fBgnunet\-download\fP(1), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/fs/tools/gnunet-directory.c
/*
     This file is part of GNUnet.
     (C) 2002, 2003, 2004, 2005, 2006 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/fs/tools/gnunet-directory.c
 * @brief tool to list the entries stored in the database holding
 *        files for building directories, to delete all of these
 *        entries and to display the contents of  directories.
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_directories.h"
#include "gnunet_ecrs_lib.h"
#include "gnunet_uritrack_lib.h"
#include "gnunet_util.h"

static char *cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;

static int do_list;

static int do_kill;

static int do_track;

static struct GNUNET_GE_Context *ectx;

static int
itemPrinter (EXTRACTOR_KeywordType type, const char *data, void *closure)
{
  printf ("\t%20s: %s\n",
          dgettext (LIBEXTRACTOR_GETTEXT_DOMAIN,
                    EXTRACTOR_getKeywordTypeAsString (type)), data);
  return GNUNET_OK;
}

static void
printMeta (const struct GNUNET_MetaData *meta)
{
  GNUNET_meta_data_get_contents (meta, &itemPrinter, NULL);
}

static int
printNode (const GNUNET_ECRS_FileInfo * fi,
           const GNUNET_HashCode * key, int isRoot, void *unused)
{
  char *string;

  string = GNUNET_ECRS_uri_to_string (fi->uri);
  printf ("%s:\n", string);
  GNUNET_free (string);
  printMeta (fi->meta);
  return GNUNET_OK;
}

static void
printDirectory (const char *filename)
{
  unsigned long long len;
  struct GNUNET_MetaData *md;
  char *data;
  int ret;
  char *name;
  int fd;

  name = GNUNET_expand_file_name (ectx, filename);
  printf (_("==> Directory `%s':\n"), name);
  if ((GNUNET_OK != GNUNET_disk_file_size (ectx, name, &len, GNUNET_YES))
      || (len == 0))
    {
      printf (_("=\tError reading directory.\n"));
      GNUNET_free (name);
      return;
    }
  md = NULL;
  fd = GNUNET_disk_file_open (ectx, name, O_LARGEFILE | O_RDONLY);
  if (fd == -1)
    {
      ret = -1;
    }
  else
    {
      data = MMAP (NULL, len, PROT_READ, MAP_SHARED, fd, 0);
      if (data == MAP_FAILED)
        {
          GNUNET_GE_LOG_STRERROR_FILE (ectx,
                                       GNUNET_GE_ERROR | GNUNET_GE_ADMIN |
                                       GNUNET_GE_BULK, "mmap", name);
          ret = -1;
        }
      else
        {
          ret =
            GNUNET_ECRS_directory_list_contents (ectx, data, len, NULL, &md,
                                                 &printNode, NULL);
          MUNMAP (data, len);
        }
      CLOSE (fd);
    }
  if (ret == -1)
    printf (_("File format error (not a GNUnet directory?)\n"));
  else
    printf (_("%d files found in directory.\n"), ret);
  if (md != NULL)
    {
      printMeta (md);
      GNUNET_meta_data_destroy (md);
    }
  printf ("\n");
  GNUNET_free (name);
}

/**
 * All gnunet-directory command line options
 */
static struct GNUNET_CommandLineOption gnunetdirectoryOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Perform directory related operations.")),     /* -h */
  {'k', "kill", NULL,
   gettext_noop
   ("remove all entries from the directory database and stop tracking URIs"),
   0, &GNUNET_getopt_configure_set_one, &do_kill},
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'l', "list", NULL,
   gettext_noop ("list entries from the directory database"),
   0, &GNUNET_getopt_configure_set_one, &do_list},
  {'t', "track", NULL,
   gettext_noop ("start tracking entries for the directory database"),
   0, &GNUNET_getopt_configure_set_one, &do_track},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_VERBOSE,
  GNUNET_COMMAND_LINE_OPTION_END,
};

int
main (int argc, char *const *argv)
{
  int i;
  struct GNUNET_GC_Configuration *cfg;

  i = GNUNET_init (argc,
                   argv,
                   "gnunet-directory [OPTIONS] [FILENAMES]",
                   &cfgFilename, gnunetdirectoryOptions, &ectx, &cfg);
  if (i == -1)
    {
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  if (do_list)
    printf (_("Listed %d matching entries.\n"),
            GNUNET_URITRACK_list (ectx, cfg, GNUNET_YES, &printNode, NULL));
  if (do_kill)
    {
      GNUNET_URITRACK_toggle_tracking (ectx, cfg, GNUNET_NO);
      GNUNET_URITRACK_clear (ectx, cfg);
    }
  if (do_track)
    GNUNET_URITRACK_toggle_tracking (ectx, cfg, GNUNET_YES);

  while (i < argc)
    printDirectory (argv[i++]);

  GNUNET_fini (ectx, cfg);
  return 0;
}

/* end of gnunet-directory.c */
